components/google-drive/changes-to-any-file-in-drive.js

```diff
@@ -0,0 +1,203 @@
+    },
+    timer: {
+      label: "Push notification renewal schedule",
+      description:
+        "The Google Drive API requires occasionaly renewal of push notification subscriptions. **This runs in the background, so you should not need to modify this schedule**.",
+      type: "$.interface.timer",
+      default: {
+        intervalSeconds: 60 * 60 * 6,
+      },
+    },
+  },
+  hooks: {
+    async activate() {
+      // Called when a componenent is created or updated. Handles all the logic
+      // for starting and stopping watch notifications tied to the desired files.
+
+      const channelID = this.db.get("channelID") || uuid();
+
+      // Subscriptions are tied to Google's resourceID, "an opaque value that
+      // identifies the watched resource". This value is included in request headers
+      let subscription = this.db.get("subscription");
+
+      const startPageToken = await this.googleDrive.getPageToken();
+      const { expiration, resourceId } = await this.googleDrive.watchDrive(
+        channelID,
+        this.http.endpoint,
+        startPageToken
+      );
+      // We use and increment the pageToken as new changes arrive, in run()
+      this.db.set("pageToken", startPageToken);
+
+      // Save metadata on the subscription so we can stop / renew later
+      this.db.set("subscription", { resourceId, expiration });
+      this.db.set("channelID", channelID);
+    },
+    async deactivate() {
+      const channelID = this.db.get("channelID");
+      if (!channelID) {
+        console.log(
+          "Channel not found, cannot stop notifications for non-existent channel"
+        );
+        return;
+      }
+
+      const { resourceId } = this.db.get("subscription");
+      if (!resourceId) {
+        console.log(
+          "No resource ID found, cannot stop notifications for non-existent resource"
+        );
+        return;
+      }
+
+      await this.googleDrive.stopNotifications(channelID, resourceId);
+
+      // Reset DB state
+      this.db.set("subscription", null);
+      this.db.set("channelID", null);
+    },
+  },
+  async run(event) {
+    // This function is polymorphic: it can be triggered as a cron job, to make sure we renew
+    // watch requests for specific files, or via HTTP request (the change payloads from Google)
+
+    let subscription = this.db.get("subscription");
+    const channelID = this.db.get("channelID");
+    const pageToken = this.db.get("pageToken");
+
+    // Component was invoked by timer
+    if (event.interval_seconds) {
+      if (!subscription || !subscription.resourceId) {
+        return;
+      }
+      console.log(
+        `Checking for resubscription on resource ${subscription.resourceId}`
+      );
+      // If the subscription for this resource will expire before the next run,
+      // stop the existing subscription and renew
+      if (subscription.expiration < +new Date() + event.interval_seconds) {
```

Comment:
`+new Date()` is epoch in milliseconds which you are adding seconds to so this seems wrong. What are the units for subscription.expiration ?

Reply:
Boops, you're right. expiration is ms so will convert `interval_seconds`